automatic channel retune

rocus opened the pull request:
It is the intention of this change that If you are using welle-cli with the webinterface you can choose any radiostation (as scanned bij welle.io) by means of an adapted m3u file. If a channel change is needed it will be done automatically. Because such a channel change takes some time your streamer will problably time out. Wait some seconds (untill after the channel change is completed , normally 10 seconds)  and choose the radiostation again.
In the m3u file for a specific radiostation the required channel is incorporated. You can craft such an m3u file with two helper files (in the root directory of this repository). You can also made them by hand by looking at the normal browser-webinterface of welle-cli. "old" m3u files still work but don't initiate a channel change.

The webinterface in a browser stll works as it used to be: you see only the radiostations of the currently chosen channel and you can change the channel by hand.

There are only a few changes in the file WebRadioInterface.cpp. Because I am not an experienced c++ programmer, maybe you can improve on some changes!

This is also my first attempt to use git and github  so please be gentle!

mpbraendli commented:
Hello and thanks for your patch!

However, your change overlaps with pull request #742. Could you maybe coordinate with @comcloudway to see how to combine these two contributions, and how to go forward with this?

In a previous discussion we decided against doing automated channel switches using the `mp3` endpoint, so in his pull request he introduced a `mp3-channel` endpoint instead.

comcloudway commented:
Hi,
given that I haven't been able to figure out how to solve #742 it might be a good idea to use a different approach to automatic re-tuning.

I haven't tested your code yet, but I'm honestly not sure if it works. On my raspberry pi 4 4GB re-tuning takes about 30 seconds, which is why I'm using a while loop to wait until the channels are available. Maybe I'm missing something, but it does not appear as if you add a delay at some point.

To address @mpbraendli 's concern, maybe you could just rename the the endpoint to use /mp3-channel and move the regex into a separate if statement.

rocus commented:
I am using welle-cli with the changes  I proposed in my pull request on a regular basis. Yes a channel change takes some time ( sometimes only 5 seconds , sometimes 15 seconds: I use a rather old PC with a rtl.sdr stick) but that is apparently inherent in a DAB channel change. The alternative for me would be a channel change by means of the website. That takes the same amount of time. During the change I can start a stream from my welle-cli PC to the streamer but the streamer just can not connect. If you are impatient you can try to connect several times and in the end your streamer connects.This trying to connect does not have advers effects on the program (as far as I can see). You can even change the channel during a channel change: in the end you get the channel you last requested.

In another issue comment of mine I tried to change the channel outside the welle-cli program by means of a small watchdog script (running in the background looking for desired channel change requests). The welle-cli program just noticed that the desired stream was not in the current channel; the watchdog script requested a channel change by means of the webinterface. This hack worked very well but was in my opinion too much of a hack.

The disadvantage of my pull request is just that it takes some time to get your radio station but that is unavoidable; the alternative is that you have a DAB stick running for each channel.

The @mpbraendli 's concern has to do with the two helper programs I introduced in the pull request and in the issue 747. They are just optional to spare you the work to make m3u's (or pls's) for your radio stations. I can receive about 40 radio stations (with a short wire as antenna) and I hate to make 40 m3u's . For even two radio stations I am prepared to automate that!

The "new" idea for these helper files is to use the configuration file of welle.io. Welle.io and welle-cli seem to be separate programs (maybe they share some code?) but you can use the output of one to make input files for the other.

The "regex" remark is about the lines I changed in the pull request 748? I remember I had two regex expressions because I wanted the old m3u file to work (ipaddress:8888/mp3/0x1057) as well as the new one (   mp3/0x1057/5D). I suppose you could code that better....

I read somewhere that an Raspberry Pi (4?) is hardly capable of serving one radio stream. That may explain why your channel change takes 30 seconds. Can a Pi serve two radiostreams at the same time? (the same or different stations)

comcloudway commented:
I'm running postmarketOS and the welle-gui on my pi and it does work properly, beforehand I used kodi and the welle-cli version from my #740 original PR.

#740 introducted a /mp3-channel/<channel>/<programme> endpoint which can be used to switch to a new channel and play a new program. So it basically served as a shorthand for requesting a retune and playing an mp3 channel.

As I said, I'd have to take a closer look at your code, because I don't understand how streaming the mp3 immediately after calling `retune` is possible, given that we all agree on the fact that retuning takes some time. ( I'm sure you implemented something to fix this, but I only took a quick look at everything)

rocus commented:
Maybe I was not clear on that but I don't send an mp3 stream after the call  to retune. That is why when you choose a "different" radio station you get a "no connection" on your streamer. So you always have to do a retry on your streamer (or PC or whatever). This approach circumvents problably the problems you had. The second time you (try to) make a connection may or maynot succeed (depending on your patience and the retune time). 
In a user manual this should be mentioned: the first time you make a stream connection may fail (because of the wrong channel), the second (or third) will succeed.

comcloudway commented:
Well, the solution I proposed in my PR #740 circumvents this exact problem, by waiting till the retuning process has finished.
That way your m3u file may contain a link to /mp3-channel/<channel>/<programme>, which you request once, which will retune and play the stream without having to request the resource twice.

The only issue is that apparently some part of the backend removes the programme handler, which causes the CLI to crash when a client disconnects.

Assuming that I might be able to solve that issue (and merge #742) the steps you have to take would be:
1. open music player (e.g. vlc)
2. request /mp3-channel/<channel>/<programme>
3. wait for retune and the audio playback will begin automatically.

Actually you can even try this out by cloning the current next branch, but welle-cli will probably crash after you disconnect the client / close the stream

comcloudway commented:
@mpbraendli I reckon we might want to unmerge #740.
Then we'll have @rocus separate the endpoint to use the `mp3-channel` syntax and merge this (edited) PR.
Given that my #740 also included the `retune` command that would allow me to reimplement #740 and #742 and file a new PR once I've solved the issue keeping #742 from being merged.

This would make the welle-cli usable again and give me some time to figure out how to keep the CLI from crashing.

Note: I think this PR should be based on the next-branch and merged into the next-branch – but I'm not sure.

rocus commented:
I totally agree that it would be nice if you only once had to select your stream. I found 2 (or more selections) acceptable also because the changes in the program are minimal and the program stays downwards compatable. I am only curious if my streamers would time out during the retuning. I think they will but maybe there are streamers who will wait longer. Or is there some sort of stream underway during a retune? Or could that be programmed?

comcloudway commented:
I think you are actually not the first person to use the same endpoint for two functions. I was mentioned in #471 and it appears that someone posted yet another version to address this issue on a rtl forum using the `/mp3` endpoint for two things. Additionally, I don't remember at which point we agreed to use separate endpoints @mpbraendli, but I think it makes more sense to separate it to make it clearer which endpoint can be used for which purpose.

Yup, one potential problem of my code is that the stream might time out given that `retune` takes multiple attempts.
So far I've never had it happen to me, but I think having a stream while waiting for welle to retune should be a feature.
From what I remember it might be hard though, given that streaming mp3 changes the response type which makes sending a `404 programme not found` impossible.

mpbraendli commented:
Hello all, I believe the best approach if you could prepare a pull request against the "next" branch, with the proposed changes: undoing the first implementation from #740 and replace by another approach.

I'll close this pull request in the meantime, but you can of course continue the discussion in preparation of the new PR here. Is that ok for you?
